### ISSUE TITLE
Large allocation slowdown

### ISSUE THREAD

**Issue by teh** [pfnet/rflex#11]
I recently traced some slowness to [this allocation](https://github.com/pfnet/rflex/blob/master/src/codegen.rs#L336):

```rust
        let mut cmap: Vec<usize> = Vec::with_capacity(0x110000);
        cmap.resize(0x110000, 0);
```

I understand that this is necessary to support unicode (max 0x10ffff), but I suspect that high code points (>0xff) are pretty rare when lexing e.g. programming languages. I sped up my lexing by roughly 100x by reducing allocation to the largest whitespace (0x202a), and then using

```rust
if zz_input < 0xff { self.cmap[zz_input as usize] } else { 0usize }
```

later for advancing the state. I wonder whether it might make sense to be smarter about this? E.g.

1. use sth like `lazy_static!` to allocate the table
2. replace the lookup table with a large `match` and let llvm sort it out
3. use `HashMap` for the sparse region > 0xff

**Comment by suma** [comment]
Thank you for creating issue. Exactly, an intention of `cmap` is to support unicode and it is huge allocation when lexing without unicode letters.
I like No. 3 method so I'll work on in a few days.

**Comment by suma** [comment]
@teh Could you check PR #12 especially performance?

**Comment by suma** [comment]
I released v0.6.0 https://crates.io/crates/rflex :)
